### PR TITLE
102 make distinction between adding mutable and immutable properties to property list compile time based

### PR DESCRIPTION
**Pull request by Gleethos**

### CODE REVIEW

File: src/main/java/sprouts/Vars.java [hunk 1]
```diff
@@ -729,14 +797,52 @@ default Vars<T> addAll( Iterable<T> items ) {
...
+                throw new IllegalArgumentException("Null values are not allowed in this property list.");
+
+            cloned.add(
+                this.allowsNull() ?
+                    Var.ofNullable(toBeAdded.type(), toBeAdded.orElseNull()) :
+                    Var.of(toBeAdded.get())
+            );
+        }
+        return addAll(cloned);
+    }
```
Comment (Mazi-S) [inline src/main/java/sprouts/Vars.java:832]: This is inconsistent with the other add `Val` methods.

Here we allow to add non-nullable `Vals` to a nullable `Vars` list.

File: src/test/groovy/sprouts/Properties_List_Spec.groovy
Comment (Mazi-S) [inline src/test/groovy/sprouts/Properties_List_Spec.groovy:1]: Perhaps we can also add similar tests that consider nullable lists.

**Review by Mazi-S — CHANGES_REQUESTED** [review]
I really like the distinction between `val` and `var`. 

The only thing I noticed is that it is not entirely clear when it comes to adding a nullable `val`.

File: src/main/java/sprouts/Vars.java [hunk 2]
```diff
@@ -252,13 +252,33 @@ default Vars<T> add( T value ) {
...
+    /**
+     * Wraps the item of the provided {@link Val} in a new {@link Var} property
+     * and adds it to the end of the property list.
+     *
+     * @param var The value to add as a property item.
+     * @return {@code this} list of properties.
+     * @throws NullPointerException if {@code null} is not allowed and the {@code var} is {@code null}.
+     * @throws NullPointerException if the {@code var} is {@code null}.
+     */
+    default Vars<T> add( Val<T> var ) {
```
[the remark this answers is not in the document]
Comment (Gleethos) [reply src/main/java/sprouts/Vars.java:278]: Good point, I will fix that

File: src/main/java/sprouts/Vars.java [hunk 3]
```diff
@@ -592,13 +616,37 @@ default Vars<T> addAt( int index, T item ) {
...
+     * @param val   the value to add as a property item.
+     * @return {@code this} list of properties.
+     * @throws NullPointerException if {@code null} is not allowed and the {@code val} is {@code null}.
+     */
+    default Vars<T> addAt( int index, Val<T> val ) {
+        if ( val.allowsNull() )
+            return addAt(index, Var.ofNullable(val.type(), val.orElseNull()));
+        else
+            return addAt(index, Var.of(val.get()));
+    }
```
[the remark this answers is not in the document]
Comment (Gleethos) [reply src/main/java/sprouts/Vars.java]: I agree! Pushed the changes. Thanks for noticing this!

File: src/test/groovy/sprouts/Properties_List_Spec.groovy
Comment (Gleethos) [reply src/test/groovy/sprouts/Properties_List_Spec.groovy:1]: I agree and added a few tests. Thanks for the suggestion.

**Comment by Gleethos** [comment]
I pushed a whole lot of changes based on your feedback. Thank you!
Let me know if it is fine.

File: src/main/java/sprouts/Vars.java (hunk 1 above)
Comment (Mazi-S) [reply src/main/java/sprouts/Vars.java]: I totally agree. We should be consistent with the other add `Val` methods.
And I would also prefer the flexible approach you describe.

**Review by Mazi-S — COMMENTED** [review]
looks good :slightly_smiling_face:

**Review by Mazi-S — APPROVED** [review]

**Pull request merged**